libnavigation-core/src/test/java/com/mapbox/navigation/core/routerefresh/RouteRefreshControllerTest.kt

```diff
@@ -1,307 +1,690 @@
-        routeRefreshOptions,
-        directionsSession,
-        tripSession,
-        threadController,
-        routeDiffProvider,
-    )
-
-    private val requestId = 1L
-
     @Before
     fun setup() {
-        every { tripSession.getRouteProgress() } returns mockk {
-            every { currentLegProgress } returns mockk {
-                every { legIndex } returns 0
-            }
-        }
-        every {
-            directionsSession.requestRouteRefresh(any(), any(), capture(routeRefreshCallbackSlot))
-        } returns requestId
         LoggerProvider.setLoggerFrontend(logger)
     }
 
     @Test
-    fun `should refresh route every 5 minutes by default`() = coroutineRule.runBlockingTest {
-        every { routeOptions.enableRefresh() } returns true
+    fun `route with disabled refresh never refreshes`() = coroutineRule.runBlockingTest {
+        val testRoute = createNavigationRoute(
+            createDirectionsRoute(
+                routeOptions = createRouteOptions(
+                    enableRefresh = false
+                )
+            )
+        )
+        val routeRefreshController = createRouteRefreshController()
 
-        routeRefreshController.restart(validRoute) {}
-        coroutineRule.testDispatcher.advanceTimeBy(TimeUnit.MINUTES.toMillis(15))
-        routeRefreshController.stop()
+        val refreshJob = async { routeRefreshController.refresh(listOf(testRoute)) }
+        advanceTimeBy(TimeUnit.HOURS.toMillis(3))
 
-        verify(exactly = 3) { directionsSession.requestRouteRefresh(any(), any(), any()) }
+        assertTrue(refreshJob.isActive)
+        refreshJob.cancel()
     }
 
     @Test
-    fun `should refresh route according to options`() = coroutineRule.runBlockingTest {
-        val routeRefreshController = RouteRefreshController(
-            RouteRefreshOptions.Builder()
-                .intervalMillis(TimeUnit.MINUTES.toMillis(1))
+    fun `route refreshes`() = coroutineRule.runBlockingTest {
+        val (initialRoute, refreshedRoute) = createTestInitialAndRefreshedTestRoutes()
+        val directionsSession = mockk<DirectionsSession>().apply {
+            onRefresh { _, _, _, callback -> callback.onRefreshReady(refreshedRoute) }
+        }
+        val routeRefreshController = createRouteRefreshController(
+            directionsSession = directionsSession,
+            routeRefreshOptions = RouteRefreshOptions.Builder()
+                .intervalMillis(30_000)
                 .build(),
-            directionsSession,
-            tripSession,
-            threadController,
         )
-        every { routeOptions.enableRefresh() } returns true
 
-        routeRefreshController.restart(validRoute) {}
-        coroutineRule.testDispatcher.advanceTimeBy(TimeUnit.MINUTES.toMillis(15))
-        routeRefreshController.stop()
+        val refreshJob = async { routeRefreshController.refresh(listOf(initialRoute)) }
+        advanceTimeBy(TimeUnit.SECONDS.toMillis(30))
```

Comment:
Do we need to advance the time if the test runs in `runBlockingTest`?

Reply:
I think yes. 

```
 @Test
    fun testAdvanceTime1() = coroutineRule.runBlockingTest { // fails
        val result = async {
            delay(50_000)
            "completed"
        }
        advanceTimeBy(49_000)
        assertEquals("completed", result.getCompletedTest())
    }

    @Test
    fun testAdvanceTime2() = coroutineRule.runBlockingTest { // passes
        val result = async {
            delay(50_000)
            "completed"
        }
        advanceTimeBy(50_000)
        assertEquals("completed", result.getCompletedTest())
    }
```